File: JS/3606.coupon_code_validator [1].js
/**
 * @param {string[]} code
 * @param {string[]} businessLine
 * @param {boolean[]} isActive
 * @return {string[]}
 */
var validateCoupons = function(code, businessLine, isActive) {
    function isValidCode(idx){
        const s = code[idx]

        if(s.length === 0)
            return false
        
        for(const ch of s){
            if(ch >= 'a' && ch <= 'z')
                continue
            if(ch >= 'A' && ch <= 'Z')
                continue
            if(ch >= '0' && ch <= '9')
                continue
            if(ch === '_')
                continue

            return false
        }

        return true
    }

    function businessLineToInt(idx){
        const validBusinessLines = {
            'electronics': 0,
            'grocery': 1,
            'pharmacy': 2,
            'restaurant': 3
        }

        for(const [key, val] of Object.entries(validBusinessLines))
            if(businessLine[idx] === key)
                return val

        return -1
    }

    const idxs = []
    for(let i=0; i<code.length; i++){
        if(!isActive[i])
            continue
        if(!isValidCode(i))
            continue
        if(businessLineToInt(i) === -1)
            continue
        
        idxs.push(i)
    }

    idxs.sort((i, j) => {
        function lexicographical_cmp(i, j){
            const s1 = code[i]
            const s2 = code[j]

            if(s1 < s2) return -1
            if(s1 > s2) return 1
            return 0
        }

        return businessLine[i] === businessLine[j] ? lexicographical_cmp(i, j) : businessLineToInt(i) - businessLineToInt(j)
    })

    const ans = []
    for(const idx of idxs)
        ans.push(code[idx])
    
    return ans
};

File: CPP [3501-4000]/3606.coupon_code_validator [1].cpp
class Solution {
private:
    bool isValidCode(string& s){
        if(s.length() == 0)
            return false;

        for(char ch: s)
        {
            if(ch >= 'a' && ch <= 'z')
                continue;
            if(ch >= 'A' && ch <= 'Z')
                continue;
            if(ch >= '0' && ch <= '9')
                continue;
            if(ch == '_')
                continue;

            return false;
        }

        return true;
    }

    bool isValidBusinessLine(string& s){
        return s == "electronics" || s == "grocery" || s == "pharmacy" || s == "restaurant";
    }

public:
    vector<string> validateCoupons(vector<string>& code, vector<string>& businessLine, vector<bool>& isActive) {
        // {code[i], businessLine[i]}
        vector<pair<string, string>> temp;

        for(int i=0; i<code.size(); i++)
        {
            if(!isActive[i])
                continue;
            if(!isValidCode(code[i]))
                continue;
            if(!isValidBusinessLine(businessLine[i]))
                continue;

            temp.push_back({code[i], businessLine[i]});
        }
        
        sort(temp.begin(), temp.end(), [](pair<string, string>& a, pair<string, string>& b){
            auto [code_a, businessLine_a] = a;
            auto [code_b, businessLine_b] = b;

            return businessLine_a == businessLine_b ? code_a < code_b : businessLine_a < businessLine_b;
        });

        vector<string> ans;
        for(auto [code, businessLine]: temp)
            ans.push_back(code);

        return ans;
    }
};